File: 2021_12_24_advent_of_code_in_fsharp/AdventOfCode/AdventOfCode2021/Day02.fs
module AdventOfCode2021.Day2

open System

type Direction =
    | Forward of int
    | Down of int
    | Up of int

let loadData (input: string) =
    input.Split Environment.NewLine
    |> Array.map
        (fun line ->
            let parts = line.Split(' ')
            let value = Int32.Parse(parts.[1])
            match parts.[0] with
            | "forward" -> Forward value
            | "down" -> Down value
            | "up" -> Up value
            | unknownDirection -> failwith $"Unknown direction '{unknownDirection}' found in data")

type Position1 = { Horizontal: int; Depth: int }

let puzzle1 (input: string) =
    let directions = loadData input
    let finalPosition =
        directions
        |> Array.fold
            (fun state direction ->
                match direction with
                | Forward value -> { state with Horizontal = state.Horizontal + value }
                | Down value -> { state with Depth = state.Depth + value }
                | Up value -> { state with Depth = state.Depth - value })
            { Horizontal = 0; Depth = 0 }
    let result = finalPosition.Horizontal * finalPosition.Depth
    result |> string

type Position2 = { Horizontal: int; Depth: int; Aim: int }

let puzzle2 (input: string) =
    let directions = loadData input
    let finalPosition =
        directions
        |> Array.fold
            (fun state direction ->
                match direction with
                | Forward value ->
                    { state with
                          Position2.Horizontal = state.Horizontal + value
                          Depth = state.Depth + (state.Aim * value) }
                | Down value -> { state with Aim = state.Aim + value }
                | Up value -> { state with Aim = state.Aim - value })
            { Horizontal = 0; Depth = 0; Aim = 0 }
    let result = finalPosition.Horizontal * finalPosition.Depth
    result |> string

File: 2022_04_23_advent_of_code_in_csharp/AdventOfCode/AdventOfCode2021/Day02.cs
namespace AdventOfCode.AdventOfCode2021;

static class Day2
{
    public record Direction;
    public record Forward(int Value) : Direction;
    public record Down(int Value) : Direction;
    public record Up(int Value) : Direction;

    public static IEnumerable<Direction> LoadData(string input) =>
         input
            .Split(Environment.NewLine)
            .Select(line =>
            {
                var parts = line.Split(' ');
                var value = Int32.Parse(parts[1]);
                return parts[0] switch
                {
                    "forward" => new Forward(value) as Direction,
                    "down" => new Down(value),
                    "up" => new Up(value),
                    var unknownDirection => throw new Exception($"Unknown direction '{unknownDirection}' found in data")
                };
            });


    record Position1(int Horizontal, int Depth);


    public static string Puzzle1(string input)
    {
        var directions = LoadData(input);
        var finalPosition = directions.Aggregate(
           new Position1(0, 0),
           (state, direction) =>
               direction switch
               {
                   Forward(var value) => state with { Horizontal = state.Horizontal + value },
                   Down(var value) => state with { Depth = state.Depth + value },
                   Up(var value) => state with { Depth = state.Depth - value },
                   _ => throw new Exception($"Unknown direction")
               });
        var result = finalPosition.Horizontal * finalPosition.Depth;
        return result.ToString();
    }

    record Position2(int Horizontal, int Depth, int Aim);

    public static string Puzzle2(string input)
    {
        var directions = LoadData(input);
        var finalPosition = directions.Aggregate(
           new Position2(0, 0, 0),
           (state, direction) =>
               direction switch
               {
                   Forward(var value) => state with
                   {
                       Horizontal = state.Horizontal + value,
                       Depth = state.Depth + (state.Aim * value)
                   },
                   Down(var value) => state with { Aim = state.Aim + value },
                   Up(var value) => state with { Aim = state.Aim - value },
                   _ => throw new Exception($"Unknown direction")
               });
        var result = finalPosition.Horizontal * finalPosition.Depth;
        return result.ToString();
    }
}